File: 1_50/13_roman_to_integer.c
int romanToInt(char* s) {
    int res = 0;
    char *p = s;
    
    if (!p) {
        return 0;
    }
    
    while(*p) {
        switch(*p) {
            case 'I':
                res += (*(p+1) == 'V' || *(p+1) == 'X') ?
                        -1 : 1;
                break;
            case 'V':
                res += 5;
                break;
            case 'X':
                res += (*(p+1) == 'L' || *(p+1) == 'C') ?
                        -10: 10;
                break;
            case 'L':
                res += 50;
                break;
            case 'C':
                res += (*(p+1) == 'D' || *(p+1) == 'M') ?
                        -100 : 100;
                break;
            case 'D':
                res += 500;
                break;
            case 'M':
                res += 1000;
                break;
        }
        p++;
    }
    return res;
}

File: 1_50/13_roman_to_integer.py
class Solution:
    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s == "":
            return 0
        
        d = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100,
            'D': 500, 'M': 1000}
        
        res = d[s[0]]
        
        # 'I' can be placed before V and X to make 4 and 9
        # check if subsequent roman is greater than previous
        
        for i in range(1, len(s)):
            if d[s[i]] > d[s[i-1]]:
                # substract twice the added previous number
                res -= 2 * d[s[i-1]]
            res += d[s[i]]
        
        return res
    
